### PR TITLE
implemented the solution described by @rgrandl

### PR DESCRIPTION
**Pull request by naivary**
I have not tested it yet but wanted get some feedback if the code already has issues that can lead to problems in the future. The issue is described in #106.

### CODE REVIEW

File: internal/tool/ssh/deploy.go [hunk 1]
```diff
@@ -87,12 +87,14 @@ func deploy(ctx context.Context, args []string) error {
 	}
 
 	// Copy the binaries to each location.
-	if err := copyBinaries(locs, dep); err != nil {
+	tmpDirs, err := copyBinaries(locs, dep)
+
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:91]: remove empty line before checking the error. here and everywhere else

we typically do:
bla, err := ...
if err != nil {
 ....
}

File: internal/tool/ssh/deploy.go [hunk 2]
```diff
@@ -131,30 +133,59 @@ func deploy(ctx context.Context, args []string) error {
 	}
 }
 
+func getTmpDirs(locs []string, depId string) (map[string]string, error) {
```
[the remark this answers is not in the document]
Comment (rgrandl) [reply internal/tool/ssh/deploy.go:136]: also, move this function at the end of the file

File: internal/tool/ssh/deploy.go [hunk 3]
```diff
@@ -131,30 +133,59 @@ func deploy(ctx context.Context, args []string) error {
...
+	tmpDirs := make(map[string]string, len(locs))
+
+	for _, loc := range locs {
+		cmd := exec.Command("ssh", loc, "mktemp", "-u")
+
+		if err := cmd.Run(); err != nil {
+			return nil, err
+		}
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:144]: we don't need this. Actually this leads to an error.  cmd.Output() already runs the command.

**Review by rgrandl — CHANGES_REQUESTED** [review]
Thanks @naivary for working on this. It looks great.

**Comment by naivary** [comment]
Thanks for the great feedback. I will try to get it done by sunday.

Cheers

File: internal/tool/ssh/impl/manager.go [hunk 4]
```diff
@@ -81,6 +81,7 @@ type manager struct {
 	mgrAddress string   // manager address
 	registry   *status.Registry
 	started    time.Time
+	tmpDirs    map[string]string
```
[the remark this answers is not in the document]
Comment (naivary) [reply internal/tool/ssh/impl/manager.go:84]: You mean the keys should be the locations and the values the tmpdir?

**Comment by naivary** [comment]
I think the naming of locs and locations is a little bit confusing. Any sugesstions for a better naming? And is it the solution you described with the comments?

Cheers.

**Comment by rgrandl** [comment]
@naivary sorry for the late reply. Can you resolve the conflicts, so I can take another look? I will prioritize this as soon as you fix the conflicts.

**Comment by naivary** [comment]
Got some problems with the naming because one parameter was removed. I'm trying to get it done as soon as possible.

**Comment by naivary** [comment]
Now the errors should be resolved. Still the naming is confusing locs vs locations and we should discuss a better proper naming. Unfortunately I hadn't not a good idea.

Cheers.

File: internal/tool/ssh/deploy.go [hunk 5]
```diff
@@ -131,30 +133,38 @@ func deploy(ctx context.Context, args []string) error {
 	}
 }
 
-// copyBinaries copies the tool and the application binary to the given set
-// of locations.
-func copyBinaries(locs []string, dep *protos.Deployment) error {
+// copyBinaries copies the tool and the application binary
+// to the given set of locations. It produces a map which
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:137]: Returns the paths to the directories where the binaries were copied, keyed by locations.

File: internal/tool/ssh/deploy.go [hunk 6]
```diff
@@ -131,30 +133,38 @@ func deploy(ctx context.Context, args []string) error {
...
+		remotePath := filepath.Join(tmpDir, filepath.Base(binary))
+
+		cmd := exec.Command("ssh", loc, "mkdir", "-p", tmpDir)
 		if err := cmd.Run(); err != nil {
-			return fmt.Errorf("unable to create deployment directory at location %s: %w\n", loc, err)
+			return nil, fmt.Errorf("unable to create deployment directory at location %s: %w\n", loc, err)
 		}
 
-		cmd = exec.Command("scp", ex, binary, loc+":"+remoteDepDir)
+		cmd = exec.Command("scp", ex, remotePath, loc+":"+tmpDir)
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:161]: this should be `binary`. we copy the binary from the local machine (so the path should be local, as specified in the deployment). Remove `remotePath`, the remote path is simply tmpDir.

File: internal/tool/ssh/deploy.go [hunk 7]
```diff
@@ -131,30 +133,38 @@ func deploy(ctx context.Context, args []string) error {
...
 		if err := cmd.Run(); err != nil {
-			return fmt.Errorf("unable to copy app binary at location %s: %w\n", loc, err)
+			return nil, fmt.Errorf("unable to copy app binary at location %s: %w\n", loc, err)
 		}
 	}
-	return nil
+
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:166]: remove empty line

File: internal/tool/ssh/deploy.go [hunk 8]
```diff
@@ -131,30 +133,38 @@ func deploy(ctx context.Context, args []string) error {
...
+		return nil, err
 	}
 
 	binary := dep.App.Binary
-	remoteDepDir := filepath.Join(os.TempDir(), dep.Id)
-	dep.App.Binary = filepath.Join(remoteDepDir, filepath.Base(dep.App.Binary))
-	for _, loc := range locs {
-		// Make an app deployment directory at each location.
-		cmd := exec.Command("ssh", loc, "mkdir", "-p", remoteDepDir)
+
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:152]: remote empty line

File: internal/tool/ssh/deploy.go [hunk 9]
```diff
@@ -228,3 +238,22 @@ func getAbsoluteFilePath(file string) (string, error) {
...
 }
+
+// getTmpDirs prints the tmp directory in which the weaver binary will
+// be stored on the remote machine. The tmp directories are produced using
+// a dry-run and are therefore unsafe. See: https://linux.die.net/man/1/mktemp
+func getTmpDirs(locs []string, depId string) (map[string]string, error) {
+	tmpDirs := make(map[string]string, len(locs))
+	for _, loc := range locs {
+		cmd := exec.Command("ssh", loc, "mktemp", "-u")
+
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:249]: remove empty line

File: internal/tool/ssh/deploy.go [hunk 10]
```diff
@@ -228,3 +238,22 @@ func getAbsoluteFilePath(file string) (string, error) {
...
+		tmpDir, err := cmd.Output()
+		if err != nil {
+			return nil, err
+		}
+
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:254]: remove empty line

File: internal/tool/ssh/deploy.go [hunk 11]
```diff
@@ -228,3 +238,22 @@ func getAbsoluteFilePath(file string) (string, error) {
...
+		tmpDirs[loc] = filepath.Join(strings.TrimSpace(string(tmpDir)), depId)
+	}
+
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:257]: remove empty line

File: internal/tool/ssh/deploy.go [hunk 12]
```diff
@@ -228,3 +238,22 @@ func getAbsoluteFilePath(file string) (string, error) {
 	}
 	return abs, nil
 }
+
+// getTmpDirs prints the tmp directory in which the weaver binary will
```
Comment (rgrandl) [inline internal/tool/ssh/deploy.go:242]: getTmpDirs returns the path to the tmp directories where the weaver binaries will be stored at each remote location.

File: internal/tool/ssh/impl/manager.go [hunk 13]
```diff
@@ -87,11 +87,14 @@ type manager struct {
 	ctx        context.Context
 	dep        *protos.Deployment
 	logger     *slog.Logger
-	locations  []string // addresses of the locations
-	mgrAddress string   // manager address
+	mgrAddress string // manager address
 	registry   *status.Registry
 	started    time.Time
 
+	// keys: addresses of the locations
```
Comment (rgrandl) [inline internal/tool/ssh/impl/manager.go:94]: managed locations, and the corresponding directories to store the weaver binaries at each location

File: remote.go [hunk 14]
```diff
@@ -55,6 +55,7 @@ func newRemoteEnv(ctx context.Context, bootstrap runtime.Bootstrap, handler conn
 	env := &remoteEnv{
 		conn: conn,
 	}
+
```
Comment (rgrandl) [inline remote.go:58]: remove

**Review by rgrandl — COMMENTED** [review]
Looks great @naivary. Can you take a final pass on my comments? Mostly nits. Also, can you rebase again?

**Review by rgrandl — COMMENTED** [review]
Looks great @naivary. Can you take a final pass on my comments? Mostly nits. Also, can you rebase again?

**Comment by naivary** [comment]
Got it done!

**Comment by rgrandl** [comment]
@naivary ,can you please run ./dev/build_and_test? We have to update the doc to mention this step, probably it's not clear. Sorry about that.

**Comment by naivary** [comment]
All stages were successfull!

**Comment by rgrandl** [comment]
> All stages were successfull!

Great. Can you commit any new local changes after you run ./dev/build_and_test and push it again?

**Comment by naivary** [comment]
There were no local changes made. I ran git add and git commit (already up to date)

**Pull request merged**